### ISSUE TITLE
CODE: Real-time Renderer for TensoRF

### ISSUE THREAD

**Issue by UPstartDeveloper** [UPstartDeveloper/NeRF-to-XR#26]
This PR is going to take over where #9 left off

The authors of [TensoRF](https://github.com/apchenstu/TensoRF) made a great contribution to this problem domain - now if we want to use it for our use case, we'll need to extend their model by letting it work with SNeRG's baking pipeline.

Note: the code for this work will actually be housed on our sister repo, [here](https://github.com/UPstartDeveloper/Flax-TensoRF?organization=UPstartDeveloper&organization=UPstartDeveloper).

**Comment by UPstartDeveloper** [comment]
At a high level, we our work seems fairly straightforward:

The `train.py`, `eval.py`, and `bake.py` scripts in SNeRG all relied on JAXNerf, which was just a version of Mildenhall et. al.'s original work,  implemented using Flax modules.

So basically, we could create something similar to what Hedman et. al. created [here](https://github.com/google-research/google-research/blob/master/snerg/nerf/models.py), just for TensoRF. I elaborate on this [here](https://github.com/apchenstu/TensoRF/issues/7#issuecomment-1086408625).

**Comment by UPstartDeveloper** [comment]
Today: so far, refactored the spherical harmonics functions from TensoRF, in the other [repo](https://github.com/UPstartDeveloper/Flax-TensoRF).

One of the big questions - does anyone know of any tools for Flax/PyTorch interoperability? So we could slim down on our work.

**Comment by UPstartDeveloper** [comment]
update: it turns you can instantiate JAX NumPy arrays from PyTorch Tensors pretty easily. 

So today I just wrote a wrapper around the PyTorch TensoRF classes, just enough it could work with the `bake.py` in `snerg`. Will test it later tonight.

If all is well, will go ahead with writing a separate, cleaner of version of this in the [Flax-TensoRF](https://github.com/UPstartDeveloper/Flax-TensoRF) repo.

**Comment by UPstartDeveloper** [comment]
^can start by looking at this FCC course on [OpenGL](https://www.youtube.com/watch?v=45MIykWJ-C4)

**Comment by UPstartDeveloper** [comment]
^mini-update: I consulted with the SNeRG author Peter Hedman for his advice, and it was very informative. Based on our convo, it seems like the hardest part of this TensoRF renderer will be an the acceleration structure for free space skipping. 

Peter's analogy was this could be the indirection grid from SNeRG or just an octree/BVH.

**Comment by UPstartDeveloper** [comment]
^mini-update: when I asked Peter, his opinion (which I agree with) is that SNeRG doesn't really work (or might not even be appropiate) for TensoRF, unless we can train a mlp using TensoRF to predict view dependence during real-time rendering.

On the other hand though, I think it's safe to say we don't even need view-dependent effects for our use case (at least for now)? We're not necessarily trying to go for photorealistic rendering, just a solid material preview for CAD-like applications.

**Comment by UPstartDeveloper** [comment]
TODO: tomorrow: study how rendering works in the actual JavaScript side of the `snerg/viewer` code

**Comment by UPstartDeveloper** [comment]
**Update: Wed Apr 20, 2022**

_I'm going to give a super-watered down, code-friendly explanation of how the SNeRG viewer works below - I hope this is beginner-friendly enough so we can start to just see what a TensoRF renderer would need to do in order to work (since we're also trying to run real-time rendering), and then brainstorm how to do those things in TensoRF-y way..._

**Pre-requisite Know-how**: before reading this explanation, it's best if you read the [original SNeRG paper](https://arxiv.org/pdf/2103.14645.pdf) first. The whole thing is very informative, but if you're pressed for time then just look at Figure 2, with its caption (if you're a visual learner like me, then basically the right half of the image visualizes what's happening in the JS code in `snerg/viewer`).

----------
Now let's begin!

**How It Works**: All the SNeRG viewer wants to do is:
1. setup a Three.js camera
2. initializes four, [`THREE.Data3DTexture`](https://threejs.org/docs/index.html#api/en/textures/Data3DTexture) objects from our saved SNeRG assets:
       1. the `rgbVolumeTexture` - aka just the diffuse colors
       2. the `featureVolumeTexture` - aka the specular features
       3. the `atlasIndexTexture` - aka the indirection grid (this is what we're most confident needs to be changed in the TensoRF renderer)
       4. the `alphaVolumeTexture` - aka, this is an indirect way of getting the volume densities
3. using a custom fragment shader and the `gDisplayMode`, we tell the GPU what colors to put on the `Material`
       1. this is where ray marching comes into play btw - might not be super relevant, but it's prolly why the variable is called the `rayMarchMaterial`
       2. depending on if we're showing specular features --> we'll also need to evaluate a small MLP to get "view-dependent effects" (e.g. when someone reflects light off their smartphone screen into your eyes on a sunny day, but then you move out of the way and their screen looks like its normal color again - _that's_ a view-dependent effect)
4. we take the material, add it to a _plane geometry_... and BOOM! We have a `Mesh` for our scene!
5. then we do still need to do a few more things for the purpose of UX - these are some of the more common parts of Three.js:
       1. e.g. setting up a render loop, to animate the FPS counter
       2. adding orbit controls
6. notably - it's not until the `loadOnFirstFrame()` function that we actually finish "filling in" the data textures from step (2).

----
Hope this helps/is understandable/correct!

**Comment by UPstartDeveloper** [comment]
_Update: Apr 24, 2022_

**Potential Pipelines for TensoRF RT-rendering** (of which I have no idea will work):

1. _Use RGB-D scans to do the 3D reconstruction:_
      1. Pro: we could probably use the output of the `train.reconstruction(args)` function, in `TensoRF/train.py`.
      2. Con 1: the literature shows this is like a research problem in itself
      3. Con 2: I don't know if this will allow us to do deferred view dependence, like in the SNeRG pipeline (therefore, this could limit interactivity in the UI/UX)
2. _"SNeRG for TensoRF"_:
     1. Explanation: we take the RGB and depth maps from the TensoRF, and alpha-composit them to get the diffuse colors during rendering (very akin to SNeRG).
     2. Pro: could leverage a lot of the existing code in SNeRG, TensoRF, and `indexFromMesh.js` (using the `DISPLAY_DIFFUSE` mode).
     3. Con: not sure how we could add in specular features (i.e. to do view dependence). 
             1. on the other hand, we might not need view dependence (the reason why is [here](https://github.com/UPstartDeveloper/NeRF-to-XR/issues/26#issuecomment-1099666343))
3. _Use GLSL_:
      1. Explained: essentially, we would just implement the TensoRF algorithm from scratch in a fragment shader - from there, just build out the renderer using three.js
      2. pro: this is the most extensible solution (if it could work)
      3. con: this is the hardest to code

**Comment by UPstartDeveloper** [comment]
^mini-update: since it's the fastest, I started prototyping a pipeline using the 2nd idea above. Currently in the process of using TensoRF in `bake.py`, and getting rid of the use of the view dependent MLP.

**Comment by UPstartDeveloper** [comment]
**Status updates over the past few days:**

1. _Refactoring_: the functions in `tensorf_baking.py` and `tensorf_bake.py` needed several refactoring updates in terms of function params
2. right now the `extract_predictions_tensorf()` function is being updated, so that we can render the sparse 3D voxel grid
3. the major challenge with this is understanding how to predict RGB values from TensoRF - it's a little different usage than what I think the authors intended, because we are using it on a regular 3d grid point in space (as opposed to having a point and viewdir)

Will keep exploring!

**Comment by UPstartDeveloper** [comment]
**Status Update: May 1st, 2022**

**A few sentences to describe this issue:**

- Part of the SNeRG pipeline involves storing the diffuse colors (aka RGB values) predicted by the model (in their case, NeRF) onto a 3D voxel grid. 
- I am trying to do this using TensoRF, but I am confused.
- Specifically, where I'm stuck is passing the rays needed to regress the RGB values onto this voxel grid (which is stored in-memory as a 4D [tensor of [H, W, D, 3]](https://github.com/UPstartDeveloper/NeRF-to-XR/blob/f15c3ab370b34d1e7163baede6996045e8913322/snerg/bake/baking.py#L27) coordinate values in worldspace). 

**Open Questions:** 

1. Does it sound like wecould reuse the rays we already generated for the image dataset to make the prediction, or would I need to generate a new set of rays for the points on the voxel grid?
2. In general, is there any way at all to predict RGB from TensoRF, given only a XYZ point in worldspace? All the places in the code usually pass a PyTorch tensor of rays.

If this does prove impossible, we can still switch to a native WebGL approach/find a way to re-architect TensoRF for deferred shading.

**Comment by UPstartDeveloper** [comment]
^mini-update: personally I'm opting for the native WebGL approach for now. It could just be my understanding, but I don't think there's a mathematically sound way to get viewdirs for the rays that point to the blocks in the voxel grid.

**Comment by UPstartDeveloper** [comment]
**Big Picture Objectives for the TensoRF Renderer**:

1. _Render_ a TensoRF at a minimum of 30 FPS, retaining PSNR of at least 40 (this will depend a lot on the acceleration structure we use).
2. _Compress_ the memory footprint, draw calls, and polygon count to be comparable to SNeRG.
3. _Organize_ the code into a sensible structure (that will probably include HTML, JS, GLSL, and CSS) - will probably need to take inspiration from the [original TensoRF repo](https://github.com/apchenstu/TensoRF).

**Comment by UPstartDeveloper** [comment]
**What I've Learned About Real-Time Rendering**:

Thankfully there have been several works done in this area (apart from SNeRG). Here is a non-exhaustive list I've been researching - so we can kinda look at the trade-offs that might help in building our own renderer:

1. [**SNeRG**](https://phog.github.io/snerg/)
       a. _Pros_: this kind of data structure does seem pretty effective at reducing the size of assets needed for RT rendering. And of course, the rendering quality is very high (compared to what rendered images from the same radiance field look like).
       b. _Cons_: does NOT work on mobile. 
       c. _Future work_: how do you create SNeRGs for other kinds of radiance fields - this a non-trivial problem, as some algorithms (e.g. TensoRF) might not lend themselves easily to deferred view dependence
2. [**Instant-ngp**](https://nvlabs.github.io/instant-ngp/)
       a. _Pros_: awesome graphical quality! lots of good additional visual effects you can do. Super fast convergence! 
       b. _Cons_: 
           1. for our use case it is not perfect - it was designed to be used as a native application, and requires an NVIDIA GPU (so web app running on mobile client is a no-no)
           2. And its not OSS (to be fair, NVIDIA does offer [research licenses](https://www.nvidia.com/en-us/research/inquiries/))
       c. _Future work_: in order to do something like Instant-NGP on the web, there's 2 questions I have:
            1. Coding: could we use [tiny-cuda-nn](https://github.com/NVlabs/tiny-cuda-nn) to take advantage of "multiresolution hash encoding" for TensoRF (similar to #29)
            2. Architecture: how to best separate out computation, so the user doesn't need to have an entire image dataset on their local machine to see the NeRF (and be mobile friendly)?
4. PlenOctrees:
       a. _Pros_: 
              1. I'm a little biased, but I do like that they at least have an OSS web viewer we could learn from. Alex Yu is even working on a [library](https://github.com/sxyu/nerfvis) that's in the same direction.
              2. it works on newer mobile devices! (e.g. not my iPhone X, but iPhone 12) 
       b. _Cons_: 
           1. for [nerfvis](https://github.com/sxyu/nerfvis) - looks like it was still intended to be used with GPU acceleration at the moment
           2. the graphical quality is not on par with what NeRFs are capable of for some reason
       c. _Future work_: maybe we just go ahead and try plugging TensoRF into nerfvis for now? Then later on, we can see how to 1) improve the graphics, 2) improve the memory footprint, at least enough so it can work on mobile

**Comment by UPstartDeveloper** [comment]
**Status Update: May 7th, 2022**

_Brief overview of code changes made in the last 72 hr:_

1. I started a new version of nerfvis, [`tensorfvis`](https://github.com/UPstartDeveloper/tensorfvis), so we could make it work better with TensoRF more quickly. Will work on it just for our specific use case for now, simultaneously with this repo (commits will go in #31). The plan is to have the library eventually support all kinds of radiance field algorithms though, and just have users implement their own custom script after installing
2. I added new functions to the `datasets` and `tensorf` modules in the `snerg.model_zoo` package, to support TensoRF visualization (with the afforementioned `tensorfvis` library).
3. For the `tensorfvis` library - currently I'm trying to call/debug/call `set_nerf()`. Through benchmarking and reading the docs, I believe we know all the right params to pass to the function for now, and I was able to at least evaluate TensoRF on a grid (for the first time)!

4. Currently debugging this error in `set_nerf`, which is thrown I believe due to an error with `svox` and its `_get_c_extension()`:
```
File "/home/ec2-user/NeRF-to-XR/snerg/tensorfvis/scene.py", line 709, in set_nerf
    grid_weights = _calculate_grid_weights(rgb_sigma[..., -1:], c2w.float(), focal_length, image_width, image_height)
  File "/home/ec2-user/NeRF-to-XR/snerg/tensorfvis/scene.py", line 670, in _calculate_grid_weights
    opts = _C.RenderOptions()
AttributeError: 'NoneType' object has no attribute 'RenderOptions'
```

**Comment by UPstartDeveloper** [comment]
**^mini-update:** it turns out the fix for this is straightforward - the TLDR is that (for some reason), it's best to install `svox` from source. This [post](https://github.com/sxyu/svox2/issues/15#issuecomment-1008596157) has additional details.

**Comment by UPstartDeveloper** [comment]
**^mini-update**: making progress, but I have a new bug to record here:

- it looks like the grid sizes used by our TensoRF model (i.e. `[290, 279, 333]`) and that in `tensorfvis` are way off
- currently what happens is that `tensorfvis` creates a 2D array of [N, 3]
- then it uses a mask to take only the grid weights that are over a certain threshold. Our grid then becomes `[0, 3]` (i.e. totally empty)
- I believed I've traced this error down to the `_calculate_grid_weights()`. And there could be 2-3 things we could do fix it:

1. see if it's an error with the `_C.grid_weight_render()` (which means looking again at `svox.csrc`)
2. I also noticed that the `sigmas` array produced by my TensoRF is all zeros. This could signal a problem with how I implemented/called the `tensorf.compute_density()`
3. see if I need to make `grid` in `tensorfvis` have the same dims as that used by our TensoRF model.

**Comment by UPstartDeveloper** [comment]
^mini-update: I believe I made some progress today, by redesigning the `tensorf.compute_density()` method to normalize the rays before using them to predict sigma. 

However there are now some new issues to work out, mainly having to do with a mismatch in the grid volume dimensions (which is in the 100,000s by default), and the sigma vector we output (which is relatively small, right now I have it set to 9 because I didn't know what else to put).

Will open up an [issue on the `nerfvis` library](https://github.com/sxyu/nerfvis/issues/5) - hopefully we just need some clarification on what these params mean, and how to set them. Nonetheless I will study the PlenOctrees paper + code to get additional insights.

**Comment by UPstartDeveloper** [comment]
^mini-update: small win today, I discovered that `tensorBase` records the coordinates of it's grid volume in `aabb` (which stands for "axis-aligned bounding box"). There's a new helper now that uses that data to get the center point of the TensoRF. Hopefully it will help us in interfacing with the `Scene` module of `nerfvis`.

Re: blockers - we have 2 main ones at the moment:

1. the sigmas arrays is still all zeros - I am starting to think I am passing the wrong XYZ values for the ray origins. Could it be something different, since our `viewdirs` are basically just a random sampling of points (equally spread out) on the surface of a sphere?
2. still trying to get a handle on the right grid dimensions to use - again, will probably look into setting this in `nerfvis.Scene.set_nerf` using TensoRF hyperparameters.

**Comment by UPstartDeveloper** [comment]
^mini-update: did maybe 5-7 additional runs of the code today. It looks like we need to set the `chunk = sh_proj_sample_count ** 2`. Will definitely verify more tomorrow!

I do believe the `reso` variable also impacts whether the `reshape()` calls in the code work or not, but just not sure if there's a way to algebraically determine what it should be yet.

**Comment by UPstartDeveloper** [comment]
^mini-update: for tomorrow, there are two problems I have attention on:

1. the `shape` problem: we need to fix the `Runtime` errors that get raised by the `nerfvis` code, which occurs due to incorrectly shaped tensors being outputted by my `eval_fn`. There's a known fix for this that involves using the `N_samples` arg in the TensoRF code - just need to test it.
2. the `sigma` problem - this refers to how the sigma array is all zeros. This issue will hopefully be easier to resolve after the `shape` problem is solved, and I understand a little more about how spherical harmonics works conceptually. Will look at this [paper](http://www.cse.chalmers.se/~uffe/xjobb/Readings/GlobalIllumination/Spherical%20Harmonic%20Lighting%20-%20the%20gritty%20details.pdf), by Robin Green of Playstation. Right now, my hunch is just that I'm using the `TensorBase.compute_valid_density` function incorrectly - specifically, the `sample_ray()` always returns a `valid_ray` tensor that is all `False`.

**Comment by UPstartDeveloper** [comment]
mini-updates:

- the `shape` problem has disappeared! turns out the `sigma` array returned by my TensoRF code had incorrect dims for `nerfvis`. I anticipate working on the `shape` problem tomorrow, albeit I doesn't necessarily I think this is fully going away just yet. I say this for a few reasons:
- This is because my current solutions tightly couples the `sh_proj_sample_count` to the `sh_deg` - I need to verify if this is conceptually correct. If not, then we should pursue a way of coding this so that the `compute_raw_rgb` function doesn't force the `sh_proj_sample_count` to have to equal `(sh_deg + 1) ** 2`, as stated above
- On a related note, we need to decide if it's conceptually ever valid to let TensoRF have multiple sampling points on the ray when it computes sigma, like it does in the current code. If not, then it's ok - we can hard-code it to output a column vector from the `compute_valid_density` function
- For now I won't do that, since it seems like the author did intend it to be ok. However, _if_ it is okay to let there be multiple sigma values per ray, then I need to review how exactly we "composit" them down into a single value, for the final render.

**Comment by UPstartDeveloper** [comment]
-mini-update: so it looks like both NeRF-SH and TensoRF have ways of resolving multiple sigma channels down to a sigma value - one uses a DNN that has a single output neuron in the last layer (for the sigma prediction), and the other (speaking specifically of Vm decomposition) uses vector-matrix operations to get a depth-map of single values
- because we're not using a DNN, nor are we interested in returning the depth map (I think), I'm _guessing_ it should be fine to just sum the multiple sigma channels for now (and just make it clear in the paper/docs we modified TensoRF in this way).

**Comment by UPstartDeveloper** [comment]
TODO: commit

TODO: address sigma problem

TODO: address decoupling `sh_proj_sample_count` from `sh_deg` - this is less of a priority

**Comment by UPstartDeveloper** [comment]
- Attempts + Results at Solving the Sigma Problem:

1) `center` -  I tried setting it to values from the example, sticking, to the default, and using the center of the aabb of TensoRF. Still no nonzero sigmas
2) `scale` - I tried leaving it at the default, and using the `.distance_scale` property of TensoRF. Still ran into the error.
3) Then I tried to see if the problem was that the ray origins and directions were set incorrectly. I stopped using the `get_center()` function from TensoRF and tried using the `get_rays()` function instead in the body of the `eval_fn`. It has not given a non-empty NeRF yet though - in addition, I think I set the camera  transform incorrectly - 

**What's next?** Found a good [blog article](https://www.scratchapixel.com/lessons/3d-basic-rendering/volume-rendering-for-developers/intro-volume-rendering), will read to hopefully get a better conceptual understanding, and brainstorm from there.


**Outstanding TODO:**
- address decoupling `sh_proj_sample_count` from `sh_deg` - this is less of a priority

**Comment by UPstartDeveloper** [comment]
mini-update: the renderer (almost) works! 

- just to test how big of an issue the `sigma` problem is, I hacked the TensoRF code to predict sigma values using `torch.rand` 
- it indeed "fixed" the issue - with the sigma array no longer all zeros, I was able to render a scene
- the good news is I can confirm it 1) runs in real-time, 2) does not require expensive GPU acceleration, and 3) runs on iPhone X!

so for now going to go back to debugging the `sigma` problem (for real)

**Comment by UPstartDeveloper** [comment]
^mini-update: previously for this sub-problem, I was thinking we could reuse the same `c2w` matrix found in the path-traced dataset we created in Blender earlier. I have to verify, but right now I think that is actually the wrong way to go about this.

Next steps: read this article from [Scratchapixel](https://www.scratchapixel.com/lessons/3d-basic-rendering/ray-tracing-generating-camera-rays/generating-camera-rays) (which TensoRF author Anpei Chen linked in the code), and see if there's a way we could generate our own rays (specifically for each `grid_chunk`)!

**Comment by UPstartDeveloper** [comment]
^mini-update: still stuck on how to make a custom `c2w` matrix

it probably wouldn't need to be as hard if I knew of a clear way to pass a 5D coordinate directly to TensoRF, which is what it seems like `nerfivs.scene.Scene.set_nerf()` is expecting.

**Next Steps** - not too sure if it will work, but I think for now we could try using TensoRF's own grid evalution code in place of what `set_nerf` does. 

**Cons**: require tons of code changes, and "tightly couples" the test dataset to the rendering process.

**Pros:** SNeRG's baking pipeline also tightly coupled the images to the baking pipeline (so at least it doesn't seem like a crazy idea).

**Comment by UPstartDeveloper** [comment]
^**mini-update**: the code for the approach ^above was fairly straightforward, so it's currently being tested out.

**BUT** - ran into an OOM. Here's a quick rundown of what's been tried so far, hopefully seeing the details will help in brainstorming ways to solve the problem:

1. decreasing the `chunk` size: this _did_ seem to make the renderer progress farther, but it still wasn't able to fully complete evaluating TensoRF on the image dataset (which is weird because it's not that different from what the `TensoRF.train` script does for rendering, which does work.
2. exporting the `GLIB` env var from inside pytorch: this is a different error, where it seemed like JAX wasn't using the GPU backend. But actually all that needed was reloading the terminal. 
3. disabling JAX preallocation - I thought this would do it, but I still ran into the out of memory. For now I am still setting `XLA_PYTHON_CLIENT_ALLOCATOR=platform`, so it's at least as low as possible (no reason to use more than we need to, right?) All the `nerfvis` and `TensoRF` computation relies on PyTorch.
4. adding a call to `torch.cuda.empty_cache()` - it was cool to see how this works, but I then learned this doesn't actually change how much memory PyTorch uses.  

**Next Steps**:
1. reached out to Alex Yu to try and get his advice.
2. Will keep digging in the [PyTorch docs](https://pytorch.org/docs/stable/notes/cuda.html#cuda-memory-management) - judging from the memory summaries from PyTorch the system has enough memory. So maybe I'm just not making enough of the "reserved" memory available .
3. If all else fails - could still try looking into building the renderer in pure WebGL!

**Comment by UPstartDeveloper** [comment]
^**mini-update**: still have the OOM. Here's a quick rundown of the new things I've tried so far, for future brainstorming:

1. decreasing the `chunk` size: I lowered it to 32, still was unsuccessful
2. using `torch.cuda.empty_cache()` after calling `del` and calling `gc.collect()` - I kinda expected this, but no change
3. I cut out the `flax` module for TensoRF entirely - so now we're directly interfacing with the TensoRF code in PyTorch (but I'm not exactly sure how much this lowered memory use - will add more logging statements)
4. I found about a way to explicitly tell PyTorch how much memory it is allowed ([docs](https://pytorch.org/docs/1.10/generated/torch.cuda.set_per_process_memory_fraction.html#torch.cuda.set_per_process_memory_fraction)). I wasn't sure what it was before, but I set it to 95% to see if it would work.

**Next Steps**:
1. maybe need to downsample the images more?
2. again if all else fails - could still try looking into building the renderer in pure WebGL!

**Comment by UPstartDeveloper** [comment]
Here's the error I get btw, saving it here to we can compare the stats it provides to what future error logs show (which come after future code changes are made):

```
CUDA out of memory. 
Tried to 
allocate 2.00 MiB (GPU 0; 
7.44 GiB total capacity; 
6.00 GiB already allocated; 
576.00 KiB free;
7.07 GiB allowed;
6.75 GiB reserved in total by PyTorch) 
If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
```

**Comment by UPstartDeveloper** [comment]
^**mini-update**: OOM, but learning quite a few things. Another quick rundown:

1. decreasing the `chunk` size: I lowered it to 4, let's see if it works
2. it turns out the `memory_fraction` variable does _not_ increase the amount of memory available to PyTorch

**Questions to Answer**:
1. what _exactly_ does the values in the output of PyTorch's [memory_summary()](https://pytorch.org/docs/stable/generated/torch.cuda.memory_summary.html) function mean? what problems could it help uncover?
2. maybe need to downsample the images more? would this require re-training TensoRF so the accuracy doesn't take a hit (I guess it doesn't really matter at this stage though, if we're just interested in seeing if we can get past the OOM at all)
3. would it help to run this CUDA application in Docker? I'm wondering now if the only way to free up memory is to isolate our PyTorch code from other CUDA kernels on the EC 2 instance (e.g. from cublas, cudnn, etc.).

**Next Steps**:
1. again if all else fails - could still try looking into building the renderer in pure WebGL!****

**Comment by UPstartDeveloper** [comment]
just another thought:
- this is not exactly an elegant solution - but if it ends up we can't tell exactly why the OOM happens in `set_nerf()` and not when running the TensoRF code, then I could just run this process in two scripts:
1. one script is just for evaluating TensoRF on the test images (maybe because it doesn't use a custom CUDA kernel, the memory will be fine?)
2. seconds script could just pick up on what `set_nerf()` does after grid evaluation, i.e. calling the `_calculate_grid_weights()` function

**Comment by UPstartDeveloper** [comment]
latest error (batch size was 4 samples, it just took longer to reach the OOM I think):

```
CUDA out of memory. Tried to allocate 2.00 MiB (GPU 0; 7.44 GiB total capacity; 6.04 GiB already allocated; 576.00 KiB free; 7.44 GiB allowed; 6.75 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
```

**Comment by UPstartDeveloper** [comment]
^mini-update:

- implemented the idea mentioned above - will try to render `rgb_sigma` from TensoRF on 1 script, then use `set_nerf()` on another
- re: the OOM, I think two variables we should watch are `args.render_path` and `test_dataset.render_path` - these weren't documented in the original repo, but I think I might have been using them incorrectly. I was passing in 4x4 matrices for `c2ws` using `test_dataset.poses`, but in the original TensoRF code it comes from `test_dataset.render_path` (based on whether `args.render_path`  is 0/1). 
- So I think it'd definitely be worth it to figure out what `test_dataset.render_path` is, and how it works (I haven't yet been able to find it in the `dataLoader` classes, but we can probably figure it out by observing the behavior of the `renderer.evaluation()` function

**Comment by UPstartDeveloper** [comment]
TODO: tomorrow - if `evaluation()` works fine - let's just try running our own `evaluator_for_rgb_sigma()` without passing c2w (b/c it looks like it's redundant for getting the rays, and it might just be what's causing the OOM

**Comment by UPstartDeveloper** [comment]
[We got a reply on what `test_dataset.render_path` means](https://github.com/apchenstu/TensoRF/issues/21#issuecomment-1139230961) - turns out it's only applicable to certain datasets. Should probably won't need to worry about it. 

> ^mini-update:
> 
> * implemented the idea mentioned above - will try to render `rgb_sigma` from TensoRF on 1 script, then use `set_nerf()` on another
> * re: the OOM, I think two variables we should watch are `args.render_path` and `test_dataset.render_path` - these weren't documented in the original repo, but I think I might have been using them incorrectly. I was passing in 4x4 matrices for `c2ws` using `test_dataset.poses`, but in the original TensoRF code it comes from `test_dataset.render_path` (based on whether `args.render_path`  is 0/1).
> * So I think it'd definitely be worth it to figure out what `test_dataset.render_path` is, and how it works (I haven't yet been able to find it in the `dataLoader` classes, but we can probably figure it out by observing the behavior of the `renderer.evaluation()` function

**Comment by UPstartDeveloper** [comment]
woah - downsampling by a factor of 10 took care of the OOM. What a simple solution 😅

**Comment by UPstartDeveloper** [comment]
going to record the co that did it:

```
python train.py --config configs/engine_sh.txt --ckpt log/tensorf_engine_VM/checkpoint_30000.th --render_only 1 --render_test 1 --render_rgb_sigma_only 1 --downsample_train 10 &
```

TODO: re-run the 2nd part - the set_nerf script:
```
python -m snerg.tensorf_plenoctree --config=configs/tensorf_render &
```

**Comment by UPstartDeveloper** [comment]
**^mini-update:**

1. Latest Error log:
```
(pytorch_p38) [ec2-user@ip-172-31-91-32 NeRF-to-XR]$ Traceback (most recent call last):
  File "/home/ec2-user/anaconda3/envs/pytorch_p38/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/ec2-user/anaconda3/envs/pytorch_p38/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/ec2-user/NeRF-to-XR/snerg/tensorf_plenoctree.py", line 185, in <module>
    app.run(main)
  File "/home/ec2-user/anaconda3/envs/pytorch_p38/lib/python3.8/site-packages/absl/app.py", line 312, in run
    _run_main(main, args)
  File "/home/ec2-user/anaconda3/envs/pytorch_p38/lib/python3.8/site-packages/absl/app.py", line 258, in _run_main
    sys.exit(main(argv))
  File "/home/ec2-user/NeRF-to-XR/snerg/tensorf_plenoctree.py", line 156, in main
    scene.set_nerf(
  File "/home/ec2-user/NeRF-to-XR/tensorfvis/tensorfvis/tensorf_scene.py", line 332, in set_nerf
    rgb_sigma = rgb_sigma[mask]
IndexError: The shape of the mask [32768] at index 0 does not match the shape of the indexed tensor [1024, 32, 4] at index 0
```
The code for this module is [here](https://github.com/UPstartDeveloper/tensorfvis/blob/new-render/tensorfvis/tensorf_scene.py).
2. Learnings today:
- today I figured out that you can explicitly choose how many sample points go on each TensoRF ray by passing `--nSamples` to `train.py`. I believe this is important, because when set this to the same value as the `reso`, I was able to bypass the error from yesterday. This was in `_calculate_grid_weights()`, when initializes `grid_data`.

3. Questions Remaining:
- why is this function failing now? I presumed `rgb_sigma` is supposed to be a 3d tensor. Are we doing something different from `Scene.set_nerf()`? - I should go back and check something like what the shape of `rgb_sigma` means, and how the sampled values should be treated.
- so far, it looks like alot of the predicted RGB + sigma values we are getting back are 0. Are they all 0? If so, could this signal an issue with our approach (and/or could we solve it by just retraining TensoRF on dowsampled imgs, or getting a bigger GPU)?

**Comment by UPstartDeveloper** [comment]
**^mini-update:** I decided to try going in an all-new direction for this.

**The Idea**
- The idea is that a PlenOctree and a TensoRF are both trying to model a 3D dense voxel feature grid, at the end of the day
- In the PlenOctree implementation, Alex Yu even included a [function](https://github.com/sxyu/svox/blob/dd6113104dd61f80729e5d37eed817292fbf5239/svox/svox.py#L867) called `from_grid()`, to create the data structure from a regular 3D grid (represented as a 4D tensor).
- However, there is _no such function_ currently in TensoRF, to create such a grid once the model is trained. Soo... I coded it up today, called `get_dense_RGBA()` :)

**Next Steps**:
1. _Retrain TensoRF with `shadingMode = "RGB"`_: for reasons (explained in the code comments in the body of `get_dense_RGBA()`), it only supports trained `TensorVMSplit` instances, which were trained with the "RGB" rendering module. So we need to retrain our model (and verify that it can still reach PSNR of at least 40). 
2. _Test Drive `get_dense_RGBA()`_: we need to verify if it can actually work without running into an OOM. If it can, then we'll pickle it.
3. _Create the PlenOctree_: thirdly, I'll modify the `Scene.set_nerf()` method in [`tensorfvis`](), so that it can load in the picked tensor. Then the final step will be to give the `from_grid()` function in the PlenOctrees extensions a go. This step will verify if that's sufficient to get the `nerfvis` rendering pipeline working as expected.

**Comment by UPstartDeveloper** [comment]
**^mini-update:** yesterday's training for `shadingMode = "RGB"` failed. Not the end of the world - I think the ec2 server probably just timed out.

Anyway, there might be a different way of looking at this. I'm not sure if it's mathematically sound, but couldn't TensoRF's RGBA predictions on the test dataset constitute the 4D tensor we need to make a PlenOctree (if we reshaped it so)?

If this is sound, the benefit is we can cast rays like normal, get better graphical quality. But it would be a lot of memory potentially, and we would have to follow a different series of experiments:

**Step 1: Path-Trace a New Dataset**: ensure the size of the test split is a 4D tensor of `[num_frames, height, width, channels = 3]`

**Step 2: Train the model like normal**

**Step 3: Build the grid**: in this paradigm, the goal is to predict on the test images like normal, then `reshape()` the stack of RGB predictions into a 4D tensor (and also `concat` to the predicted alpha values). This would require adding a new function, that's a cross between the `get_dense_RGBA()` and `TensoRF.renderer.evaluation()`.

**Step 4 and 5: Build the PlenOctree, Render**: same as yesterday, we would then use the `from_grid()` function from `svox` and be on our way to modifying `nerfvis`. If the final render looks good, that's an indication we have something working.

**Comment by UPstartDeveloper** [comment]
^TODO: if the about idea works, then we can delete `get_dense_RGBA()`

**Comment by UPstartDeveloper** [comment]
^mini-update: this is working so far, it just takes forever. Will probably downsample just to see if it can work at all. And, maybe the hashing technique from instant ngp can help?

**Comment by UPstartDeveloper** [comment]
**goal**: come up with at least 3 ideas for the architecture of a TensoRF render that is:

1. real-time
2. in the web browser
3. in color
4. mobile friendly

TODO: 
- answer: what made PlenOctrees work on mobile (perhaps it was volrend)?
- review: previous ideas
- investigate: could getDenseAlpha help in a TensoRF baking pipeline

**Comment by UPstartDeveloper** [comment]
> **^mini-update:** we're switching to NeRF pl w/ the goal of having a PSNR of 35-40, then will revisit creating a "deferred" TensoRF network
> 
> * basically, we would train like normal
> * then, store pre-computed appearance feature vectors into the XYZ coords of a sparse 3D voxel grid
> * then at rendering time, pass the viewdir and APV to the final decoding FCN in WebGL, to get the view-dependent RGB colors

the big question with this is: how would we achieve such a high PSNR?

Going to record the PSNRs I've seen so far (this is for NeRF-SH). Here are the TensorBoard visualizations of the experi

1. [Experiment 1 - 1000 iterations](https://tensorboard.dev/experiment/aOGVhYPLRZqdNq1YFgWy3A/#scalars)
1. [Experiment 2 - 8750 iterations](https://tensorboard.dev/experiment/tQKPInX8TbSmRumAZmEJYA/#scalars)

**Comment by UPstartDeveloper** [comment]
**NEW IDEA**:
- so it seems like whether we go with an indirection grid or PlenOctree - the big challenge with getting the renderer to work will be rendering out TensoRF on a 3D voxel grid
- this is probably just something I am new to, but to implement this correctly I'll have to see if there's a way to render rays w/ TensoRF without passing a `rays_chunk` - aka, to predict RGB and sigma using only XYZ points
- part of the answer (in a SNeRG baking pipeline) - would be using two trained TensoRF. One TensoRF would take the place of the per-sample MLP (would probably shade using the `RGBRender`, which only has a 3D appearance feature vector); and the other would use the MLP shading mode, so we could add in view dependence in the WebGL application

**TODO**:
- ask the question on casting rays onto a voxel grid (specifically, how to set the viewdir) on Stack Overflow
- keep reading papers on ^this as well
- finish training the RGB-TensoRF (started today)

**Comment by UPstartDeveloper** [comment]
^mini-update: the formula above didn't quite work (I tried running again using the scripts for the Tensorfvis script defined in #16). Not sure if it's because the grid points aren't actually going in the AABB, or what yet. Will investigate more tomorrow.

Afterwards, a known problem that needs solving is making sure the RGBs and sigma values returned by the `eval_fn()` in our script are both 2D array-like objects.

The highest "learning-potential" question here is still the math formula - does anyone have suggestions?
my thinking is that we could either 1) see if TensoRF is fast enough to run without any spatial strcutres, and we just do everything the paper says in native WebGL

OR 

2) we try to see if there's a way to store diffuse color predictions offline in a spatial structure, then add view dependence back in WebGL

**Comment by UPstartDeveloper** [comment]
Let's try and reconcile the differences in Dr. Hedman's [advice](https://github.com/UPstartDeveloper/NeRF-to-XR/issues/26#issuecomment-1097463274) and Dr. Chen's [advice](https://github.com/apchenstu/TensoRF/issues/7#issuecomment-1082516423):
- Chen says it's easy, just using TRF w/ OpenGL OR CUDA (from a high-level perspective)
- Hedman says it's also easier to go WebGL (vs. trying to get TRF to do deferred view dependence) - but he also brings up that the implementation (not the high-level theory) will need some kind of acceleration structure. But just what would that be?

**Comment by UPstartDeveloper** [comment]
...Similar to my [comment from April](https://github.com/UPstartDeveloper/NeRF-to-XR/issues/26#issuecomment-1104448586), let me share what I think a high-level, native WebGL-based renderer for TensoRF would look like. Hopefully this will outline clear, actionable TODOs for this project:

**The "Flow" of the WebGL Renderer**:

_TL;DR_ - just save the the decoding function for TRF, the appearance feature vectors per each XYZ coordinate, and fill in the rest in real-time so we can regress the view-dependent RGB colors.

_Long-form_
1. Reconstruct a trained TensoRF in memory
     1. what data structure we use will have to be determined by function - right now I think we might just need _1) JSON file --> creates a data texture of the **mlp weights**_ (used in the decoding function of our trained TRF); and _2) **a pickle file** of the appearance feature vectors (AFVs) over **all** the XYZ locations in the scene's AABB._
     2. as a side note, I've half a mind to also include a `alphaVolumeTexture`, for which we might choose to leverage the `getDenseAlpha()` function in TensoRF
3. Setup the Three.js Scenegraph, (i.e. add a `PlaneBufferGeometry`, orbit controls, a camera, a renderer object, etc.)
4. Create a 3D data texture for the view-dependent RGB colors
     1. this is where the shaders would probably come in - 
     3. in order to get `valid_rgbs` from the decoder function, we will need to pass 1) the XYZ points, 2) the viewing direction, and 3) the AFV. 
     4. Currently I'm really hoping the first two can be done in real-time, and for the third we can use some kind of data structure (maybe an indirection grid? or perhaps the `AlphaGridMask`?) to efficiently get the appropriate AFV
6. Add the texture to a material, material to the plane geometry, and 💥 _BAM_ 💥 - create the Mesh!
7. Start the render loop and FPS counter, like in the SNeRG viewer

**Known Risks of this Approach**: _- aka, reasons why I might rethink this idea (listed in order of the flow, tagged by level of risk)_
1. [LOW] How to save/load PyTorch tensors in JS?
     1. **For the MLP weights** - will go re-read the code in `export_snerg_scene`
     2. For the **pickle file** of the AFV - use this [library](https://stackoverflow.com/questions/57798609/javascript-can-load-python-pickle-dumped-string-but-not-file)?
1. [LOW] I'm not sure if the XYZ coordinates we query in the grid of AFVs are guaranteed to be valid, and return anything > 0 
    1. my idea on this - maybe we can try to do something similar to `getDenseAlpha()`, but just call it `getDenseFeatureGrid()`. By passing the same arg for `gridSize` to both, we might be able to resolve this doubt.  
1. [HIGH] Will need to experiment with how to implement the `MLPRender_Fea.forward` in code - **will probably need to highly leverage the existing SNeRG viewer and code in `tensorBase.py`**
1. [HIGH] I'm not sure how exactly to repurpose the indirection grid for our use case, if it comes to that - **will need to re-read the SNeRG paper and code**

**Comment by UPstartDeveloper** [comment]
^one noticeable difference between this approach and that in SNeRG is that I don't mention pre-computing the diffuse RGB colors at all. For now, I'm blindly presuming we'll still be able to reach RT rendering because 1) TRF is pretty fast and 2) Dr. Hedman seemed to think the only important bit of this is the acceleration structure.

**Comment by UPstartDeveloper** [comment]
**Next Step** - following the list of questions above, my next goal will be:

- to see if there's a way to create the assets for a TensoRF WebGL renderer. Will do this by reusing alot of the baking code in SNeRG
- the only challenges I'll focus on are: rewriting the export code to save the TensoRF weights, and the baking code to only save latent features in the texture atlas (see [above](https://github.com/UPstartDeveloper/NeRF-to-XR/issues/26#issuecomment-1153295980) for specific functions to look at)

**Comment by UPstartDeveloper** [comment]
^mini-update: started and nearly done with the new baking script - ironically I didn't get to the two functions mentioned above, but just about all the pieces "around" that code is there :)
